### PR TITLE
chore(linux): Build with meson instead of autotools 🏗️

### LINKED ISSUE #5981: chore(linux): convert ibus-keyman to build with meson

### PR DESCRIPTION
**Pull request by ermshiperete**
This change introduces meson for building ibus-keyman instead of autotools.

Closes #5981.

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by mcdurdin** [comment]
The changes to core are going to be a little bit fun to merge with epic-ldml!

File: core/src/meson.build [hunk 1]
```diff
@@ -84,7 +73,30 @@ lib = library('kmnkbp0',
...
+  'state.cpp',
+  'jsonpp.cpp',
   'utfcodec.cpp',
+)
+
+lib = library('kmnkbp0',
+  apifiles,
+  corefiles,
+  kmxfiles,
+  'mock/mock_processor.cpp',
```
Comment (mcdurdin) [inline core/src/meson.build:99]: Let's do the same with mock_files as with the others

File: core/src/meson.build [hunk 2]
```diff
@@ -58,18 +58,7 @@ if compiler.get_id() == 'emscripten'
...
-  'state.cpp',
-  'km_kbp_context_api.cpp',
-  'km_kbp_keyboard_api.cpp',
-  'km_kbp_options_api.cpp',
-  'km_kbp_state_api.cpp',
-  'km_kbp_debug_api.cpp',
-  'km_kbp_processevent_api.cpp',
-  'jsonpp.cpp',
-  'mock/mock_processor.cpp',
+kmxfiles = files(
```
Comment (mcdurdin) [inline core/src/meson.build:61]:
```suggestion
kmx_files = files(
```

and so on?

File: linux/ibus-keyman/meson.build [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+gtk =         dependency('gtk+-3.0',             version: '>= 2.4')
+x11 =         dependency('x11',                  version: '>= 1.6')
+json_glib =   dependency('json-glib-1.0',        version: '>= 1.0')
+kmnkbp =      dependency('keyman_kmn_processor', version: '>= 0.0.0')
+
+env = find_program('env')
+
+# define some variables. We wouldn't need this if we'd have a meson.build file
+# in the Keyman root directory...
+coredir = '@0@/../../core'.format(meson.current_source_dir())
```
Comment (mcdurdin) [inline linux/ibus-keyman/meson.build:22]: The problem with putting a meson.build in the root is it impacts all paths, doesn't it?

That said, I'm not all that excited about linux builds listing all the files in core separately -- it undoes a lot of the good of using meson in the first place. I'd prefer to see us building core as a static library and importing it that way.

File: linux/ibus-keyman/meson.build [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  join_paths(coredir, 'src/jsonpp.cpp'),
+  join_paths(coredir, 'src/utfcodec.cpp'),
+)
+coretestfiles = files(
+  join_paths(coredir, 'tests/unit/kmx/kmx_test_source.cpp'),
+)
+
+# Check if we have patched ibus
+code = '''#include <ibus.h>
+int x = IBUS_PREFILTER_MASK | IBUS_CAP_PREFILTER;
```
Comment (mcdurdin) [inline linux/ibus-keyman/meson.build:53]: Can we add the PR reference in the comment?

File: linux/ibus-keyman/src/meson.build [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+engine_sources = files(
+  'main.c',
+  'engine.c',
+  'engine.h',
+  'keycodes.h',
+  'keyman-service.c',
+  'keyman-service.h',
+)
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/meson.build:15]:
```suggestion
util_files = files(
  'keymanutil.c',
  'keymanutil.h',
  'kmpdetails.c',
  'kmpdetails.h',
)

engine_files = files(
  'main.c',
  'engine.c',
  'engine.h',
  'keycodes.h',
  'keyman-service.c',
  'keyman-service.h',
)
```

We should use either `_files` or `_sources` consistently.

File: linux/ibus-keyman/tests/meson.build [hunk 6]
```diff
@@ -0,0 +1,152 @@
+sources = [
```
Comment (mcdurdin) [inline linux/ibus-keyman/tests/meson.build:1]:
```suggestion
test_files = [
```

File: linux/ibus-keyman/tests/scripts/find-tests.sh [hunk 7]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+cd "$1" || exit
+while IFS= read -r -d '' file; do
+  testname=$(basename "$file" .kmx)
+  printf "$(basename "$file")\t${testname#k_}\n"
+done < <(find . -name \*.kmx -print0 | sort -z)
```
Comment (mcdurdin) [inline linux/ibus-keyman/tests/scripts/find-tests.sh:6]: Is this meant to be `< <`?

**Review by mcdurdin — CHANGES_REQUESTED** [review]
This is looking good, but I think we need to look at how we are using meson throughout the entire repo and avoid referencing files from other projects (i.e. core from linux in this case).

File: linux/ibus-keyman/meson.build (hunk 3 above)
Comment (ermshiperete) [reply linux/ibus-keyman/meson.build]: Good point. Refactored to create a static library and get rid of the direct references of core file.

File: linux/ibus-keyman/meson.build (hunk 4 above)
Comment (ermshiperete) [reply linux/ibus-keyman/meson.build]: Done

**Comment by mcdurdin** [comment]
Latest Developer build failed with:

```
[23:18:47][Step 1/4] 	copy C:\BuildAgent\work\7ac43416c45637e9\keyman\core\build\x86\Release\src\kmnkbp0-0.dll C:\BuildAgent\work\7ac43416c45637e9\keyman\developer\bin
[23:18:47][Step 1/4] The system cannot find the file specified.
```

File: linux/ibus-keyman/meson.build (hunk 3 above)
Comment (mcdurdin) [reply linux/ibus-keyman/meson.build]: For Core, we need to generate both static library and DLL on Windows -- static library used by Keyman Engine, DLL used by Developer.

**Review by mcdurdin — CHANGES_REQUESTED** [review]
This looks much closer, good work. We just need to avoid changes to Core impacting other platforms

File: core/src/meson.build [hunk 8]
```diff
@@ -51,25 +51,17 @@ if compiler.get_id() == 'msvc'
   version_res += import('windows').compile_resources('version.rc', args:['/n','/c65001'])
 endif
 
+build_static_lib = true
```
[the remark this answers is not in the document]
Comment (ermshiperete) [reply core/src/meson.build:54]: Done

File: linux/ibus-keyman/meson.build [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  conf.set('IBUS_HAS_PREFILTER', 1)
+endif
+
+conf.set('HAVE_CONFIG_H', 1)
+configure_file(output : 'config.h',
+               configuration : conf)
+
+core_include_dirs = [
+  include_directories('../../core/src'),
+  include_directories('../../core/src/kmx'),
```
Comment (mcdurdin) [inline linux/ibus-keyman/meson.build:38]:
```suggestion
  include_directories(core_dir / 'src'),
  include_directories(core_dir / 'src/kmx'),
```

File: linux/ibus-keyman/meson.build [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  include_directories('../../common/include'),
```
Comment (mcdurdin) [inline linux/ibus-keyman/meson.build:39]:
```suggestion
  include_directories(common_dir / 'include'),
```

File: linux/ibus-keyman/src/meson.build [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+include_dirs = [
+  core_include_dirs,
+  include_directories('..'),
+  include_directories('@0@/..'.format(meson.current_build_dir())),
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/meson.build:20]:
```suggestion
  include_directories(meson.current_build_dir() / '..'),
```

File: linux/ibus-keyman/src/test/meson.build [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  'keymanutil_tests.c',
+  util_files,
+]
+
+keymanutil_deps = [ibus, gtk, x11, json_glib, kmnkbp_lib]
+
+test_include_dirs = [
+  include_dirs,
+  include_directories('..'),
+  include_directories('@0@/..'.format(meson.current_build_dir())),
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/test/meson.build:11]:
```suggestion
  include_directories(meson.current_build_dir() / '..'),
```

File: linux/ibus-keyman/src/test/meson.build [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  executable(
+    'print_kmpdetails',
+    sources: [
+      'print_kmpdetails.c',
+      '../kmpdetails.c'
+    ],
+    dependencies: [ json_glib ],
+    include_directories: test_include_dirs
+  ),
+  args: [ '@0@'.format(meson.current_source_dir()) ],
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/test/meson.build:42]:
```suggestion
  args: [ meson.current_source_dir() ],
```

File: linux/ibus-keyman/src/test/meson.build [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+  'print-kmp-test',
+  executable(
+    'print_kmp',
+    sources: [
+      'print_kmp.c',
+    ],
+    dependencies: [ json_glib ],
+    include_directories: test_include_dirs
+  ),
+  args: [ join_paths('@0@'.format(meson.current_source_dir()), 'kmp.json') ],
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/test/meson.build:60]:
```suggestion
  args: [ meson.current_source_dir() / 'kmp.json' ],
```

File: linux/ibus-keyman/src/test/meson.build [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+  include_directories('@0@/..'.format(meson.current_build_dir())),
+]
+
+executable(
+  'keymanutil-tests',
+  sources: keymanutil_sources,
+  dependencies: keymanutil_deps,
+  include_directories : test_include_dirs
+)
+
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/test/meson.build:20]:
```suggestion

test_env = [
  'G_TEST_SRCDIR=' + meson.current_source_dir(),
  'G_TEST_BUILDDIR=' + meson.current_build_dir(),
]

```

Adding a variable for the repeated env; also, I think in things like this, string concatenation is easier to read.

File: linux/ibus-keyman/src/test/meson.build [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+test(
+  'keymanutil-tests',
+  find_program('run-tests.sh'),
+  env: [
+    'G_TEST_SRCDIR=@0@'.format(meson.current_source_dir()),
+    'G_TEST_BUILDDIR=@0@'.format(meson.current_build_dir()),
+  ],
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/test/meson.build:27]:
```suggestion
  env: test_env,
```

File: linux/ibus-keyman/src/test/meson.build [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+      '../kmpdetails.c'
+    ],
+    dependencies: [ json_glib ],
+    include_directories: test_include_dirs
+  ),
+  args: [ '@0@'.format(meson.current_source_dir()) ],
+  env: [
+    'G_TEST_SRCDIR=@0@'.format(meson.current_source_dir()),
+    'G_TEST_BUILDDIR=@0@'.format(meson.current_build_dir()),
+  ],
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/test/meson.build:46]:
```suggestion
  env: test_env,
```

File: linux/ibus-keyman/src/test/meson.build [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+      'print_kmp.c',
+    ],
+    dependencies: [ json_glib ],
+    include_directories: test_include_dirs
+  ),
+  args: [ join_paths('@0@'.format(meson.current_source_dir()), 'kmp.json') ],
+  env: [
+    'G_TEST_SRCDIR=@0@'.format(meson.current_source_dir()),
+    'G_TEST_BUILDDIR=@0@'.format(meson.current_build_dir()),
+  ],
```
Comment (mcdurdin) [inline linux/ibus-keyman/src/test/meson.build:64]:
```suggestion
  env: test_env,
```

File: linux/ibus-keyman/tests/meson.build [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+  'testfixture.cpp',
+  'testmodule.c',
+  'testmodule.h',
+]
+
+kmnkbp_tests_lib = cc.find_library(
+  'libkmnkbp-tests',
+  dirs: [
+    '@0@/../../../core/build/arch/debug/tests/unit/kmx'.format(meson.current_source_dir()),
+    '@0@/../../../core/build/arch/release/tests/unit/kmx'.format(meson.current_source_dir())
```
Comment (mcdurdin) [inline linux/ibus-keyman/tests/meson.build:13]:
```suggestion
    core_dir / 'build/arch/debug/tests/unit/kmx',
    core_dir / 'build/arch/release/tests/unit/kmx'
```

Same, conditionalize on whether or not we are doing a debug build?

File: linux/ibus-keyman/tests/meson.build [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+  ]
+)
+
+deps = [ibus, gtk, x11, json_glib, kmnkbp_lib, kmnkbp_tests_lib]
+
+test_include_dirs = [
+  core_include_dirs,
+  include_directories('../../../core/tests/unit/kmx'),
```
Comment (mcdurdin) [inline linux/ibus-keyman/tests/meson.build:21]:
```suggestion
  include_directories(core_dir / 'tests/unit/kmx'),
```

File: linux/ibus-keyman/tests/meson.build [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+  include_directories: test_include_dirs,
+)
+
+env_file = '/tmp/env.txt'
+pid_file = '/tmp/ibus-keyman-test-pids'
+
+setup_tests = find_program('setup-tests.sh', dirs: [join_paths(meson.current_source_dir(), 'scripts')])
+teardown_tests = find_program('teardown-tests.sh', dirs: [join_paths(meson.current_source_dir(), 'scripts')])
+run_test = find_program('run-single-test.sh', dirs: [join_paths(meson.current_source_dir(), 'scripts')])
+find_tests = find_program('find-tests.sh', dirs: [join_paths(meson.current_source_dir(), 'scripts')])
```
Comment (mcdurdin) [inline linux/ibus-keyman/tests/meson.build:39]:
```suggestion
setup_tests = find_program('setup-tests.sh', dirs: [meson.current_source_dir() / 'scripts'])
teardown_tests = find_program('teardown-tests.sh', dirs: [meson.current_source_dir() / 'scripts'])
run_test = find_program('run-single-test.sh', dirs: [meson.current_source_dir() / 'scripts'])
find_tests = find_program('find-tests.sh', dirs: [meson.current_source_dir() / 'scripts'])
```

File: linux/ibus-keyman/tests/meson.build [hunk 22]
```diff
@@ -0,0 +1,158 @@
+test_files = [
+  'ibusimcontext.c',
+  'ibusimcontext.h',
+  'testfixture.cpp',
+  'testmodule.c',
+  'testmodule.h',
+]
+
```
Comment (mcdurdin) [inline linux/ibus-keyman/tests/meson.build:8]:
```suggestion

test_env = [
  'G_TEST_SRCDIR=' + meson.current_source_dir(),
  'G_TEST_BUILDDIR=' + meson.current_build_dir(),
]

```

File: linux/ibus-keyman/tests/meson.build [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+)
+
+test(
+  'setup-wayland',
+  setup_tests,
+  args: ['--wayland', env_file, pid_file],
+  env: [
+    'G_TEST_SRCDIR=@0@'.format(meson.current_source_dir()),
+    'G_TEST_BUILDDIR=@0@'.format(meson.current_build_dir()),
+  ],
```
Comment (mcdurdin) [inline linux/ibus-keyman/tests/meson.build:61]:
```suggestion
  env: test_env,
```

File: linux/ibus-keyman/tests/meson.build [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+    protocol: 'tap',
+  )
+  test(
+    'X11-' + testname + '__no-surrounding-text',
+    run_test,
+    args: [ '--x11', '--no-surrounding-text', test_args],
+    env: [
+      'G_TEST_SRCDIR=@0@'.format(meson.current_source_dir()),
+      'G_TEST_BUILDDIR=@0@'.format(meson.current_build_dir()),
+    ],
```
Comment (mcdurdin) [inline linux/ibus-keyman/tests/meson.build:123]:
```suggestion
    env: test_env,
```

File: linux/ibus-keyman/tests/meson.build [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+    protocol: 'tap',
+  )
+  test(
+    'Wayland-' + testname + '__surrounding-text',
+    run_test,
+    args: [ '--wayland', '--surrounding-text', test_args],
+    env: [
+      'G_TEST_SRCDIR=@0@'.format(meson.current_source_dir()),
+      'G_TEST_BUILDDIR=@0@'.format(meson.current_build_dir()),
+    ],
```
Comment (mcdurdin) [inline linux/ibus-keyman/tests/meson.build:137]:
```suggestion
    env: test_env,
```

File: linux/ibus-keyman/tests/meson.build [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+    protocol: 'tap',
+  )
+  test(
+    'Wayland-' + testname + '__no-surrounding-text',
+    run_test,
+    args: [ '--wayland', '--no-surrounding-text', test_args],
+    env: [
+      'G_TEST_SRCDIR=@0@'.format(meson.current_source_dir()),
+      'G_TEST_BUILDDIR=@0@'.format(meson.current_build_dir()),
+    ],
```
Comment (mcdurdin) [inline linux/ibus-keyman/tests/meson.build:151]:
```suggestion
    env: test_env,
```

File: linux/ibus-keyman/meson.build [hunk 27]
```diff
@@ -0,0 +1,44 @@
+project('ibus-keyman', 'c', 'cpp',
+        version: run_command('cat', '../../VERSION.md', check: true).stdout().strip(),
+        license: 'GPL-2+',
+        # default_options : ['buildtype=release',
+        #                    'cpp_std=c++14',
+        #                    'b_vscrt=static_from_buildtype',
+        #                    'warning_level=2'],
+        meson_version: '>=0.53.0')
```
Comment (mcdurdin) [inline linux/ibus-keyman/meson.build:8]: linux/control gives >= 0.45. I would love to see 0.53+ or even higher?

**Review by mcdurdin — CHANGES_REQUESTED** [review]
What is our minimum meson version requirement for Linux now? I am seeing conflicting data. We should make it consistent across all meson.build project specs and in our docs.

If we can go 0.49 or higher, then we can use `/` instead of `join_paths()` which would be superfantastic.

Apart from that, requesting lots of little tidy-up changes, mostly!

File: linux/ibus-keyman/meson.build (hunk 27 above)
Comment (ermshiperete) [reply linux/ibus-keyman/meson.build:8]: Yep, oversight. But it was changed in #7882 which has been merged in the meantime.

File: linux/ibus-keyman/meson.build (hunk 9 above)
Comment (ermshiperete) [reply linux/ibus-keyman/meson.build:38]: Unfortunately this doesn't work - meson complains about absolute paths:

```
ERROR: Tried to form an absolute path to a source dir. You should not do that but use relative paths instead.
```

Comment (mcdurdin) [reply linux/ibus-keyman/meson.build:38]: Great work meson. We can I guess make a "core_source_dir" which is all dotty but it'll have to be defined for each meson.build file. Sometimes it is a little too opinionated for its own good!

Comment (mcdurdin) [reply linux/ibus-keyman/meson.build:38]: We'll still get the benefit elsewhere of using `core_dir`, just not for `include_directories` I think?

File: linux/ibus-keyman/tests/meson.build (hunk 20 above)
Comment (ermshiperete) [reply linux/ibus-keyman/tests/meson.build:21]: Unfortunately this doesn't work - meson complains about absolute paths:

```
ERROR: Tried to form an absolute path to a source dir. You should not do that but use relative paths instead.
```

**Review by mcdurdin — APPROVED** [review]
LGTM

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.43-alpha](https://keyman.com/downloads/releases/alpha/17.0.43)

**Pull request merged**